build: add clang-tidy configuration file

kylo252 opened the pull request:
- add clang-tidy configuration file
- enable additional useful checks through `clangd` that are nice to have (naming-conventions, etc)

references:
   - https://github.com/googleapis/google-cloud-cpp/blob/main/.clang-tidy
   - https://wiki.sei.cmu.edu/confluence/display/c/Clang

---

Related: #10632, #11236, #15428

kylo252 committed:
build: add clang-tidy integration

add a clang-tidy target
  ```shell
  make clang-tidy
  ```
run with a specific file
  ```shell
  make clang-tidy FILE=src/nvim/screen.c
  ```

justinmk commented on Makefile:
```diff
@@ -190,6 +190,13 @@ clint: build/.ran-cmake
 	+$(BUILD_CMD) -C build clint-full
 
+clang-tidy: generate-compilation-db
+	bash scripts/run_clang_tidy.sh $(FILE)
```
is the run_clang_tidy.sh script really needed? it's not a pattern we usually follow. Better not to spread things out if possible, also keeps the Makefile more portable (e.g. on Windows)

kylo252 replied:
Only if you want a way to run it per-file from the CLI. If you're running `clangd` then it can also manage `clang-tidy` for you and dynamically display any warnings and code-actions through LSP.

Alternatively, we can introduce another target called  `clang-tidy-file` or something, which would allow us to skip validating the variable inside make.

I didn't want to do conditional run from inside make, because that would get hacky very quickly: https://stackoverflow.com/a/45003119

dundargoc replied:
I think this could be done without much hassle inside the makefile. Something like the following should work:

```
FILE ?= $(shell git ls-files *.c)
clang-tidy: generate-compilation-db
	  clang-tidy --config-file .clang-tidy -p build $(FILE)
```

So if it's run as `make clang-tidy` then all files will be used. If it's run with `make clang-tidy FILE=src/nvim/screen.c` then it's only used on `screen.c`.

justinmk commented on Makefile:
```diff
@@ -190,6 +190,13 @@ clint: build/.ran-cmake
+
+generate-compilation-db: build/.ran-cmake
+	ninja -C build -t compdb C_COMPILER__nvim_$(CMAKE_BUILD_TYPE) > build/compile_commands.json	
+	ln -sf build/compile_commands.json .
```
looks useful 👀

kylo252 replied:
Yeah, it's the same one I mentioned in #10632

This is essential if we want to run this in the CI and not have it analyze third-party and other non-essential files. 
Let me know if you feel that it's missing some files. You can verify the list quickly with `jq`.

```bash
# you can change the filter or drop it all together
cat compile_commands.json | jq -c '.[] | select( .file | contains("src"))' | wc -l
```

The-personified-devil replied:
Is there a particular reason for creating a symbolic link to `compile_commands.json` in the base directory?
Clangd for example is already configured to search for it in `build/` through `.clangd`.

kylo252 replied:
My bad, I didn't notice it. 

It's just the standard way to let the other tools find it without hard-coding the build path. I wanted to use $(BUILD_DIR) but it's not defined.

```make
ln -sf $(BUILD_DIR)/compile_commands.json .
```

There are a few ways for cmake to use it:
1) your typical 
```shell
cd $BUILD_DIR && cmake .. 
```
2) by setting the `CMAKE_BINARY_DIR=..` variable
2) my favorite, but I believe it's relatively new so I need to check which is the oldest supported cmake version
```shell
cmake -B $BUILD_DIR
```

dundargoc reviewed (changes requested):
:+1:

justinmk commented on src/nvim/screen.c:
```diff
@@ -1882,7 +1886,9 @@ static int line_putchar(LineState *s, schar_T *dest, int maxcells, bool rl)
-      int pc, pc1, nc;
+      int pc;
+      int pc1;
+      int nc;
```
afaik we decided against this kind of change in the uncrustify discussions.

gpanders replied:
what about multiple initializations on one line?

```c
int pc = 0, pc1 = 1, nc = 3;
```

IMO these are extremely error prone and ought to be split up. If it's the same rule that governs both then I think we should not do this (but if it's possible to behave differently for initialized and uninitialized declarations then I have no objection).

justinmk replied:
i'm ok either way. @neovim/core +1 / -1

kylo252 replied:
@gpanders, I don't think the check can make that distinction: https://clang.llvm.org/extra/clang-tidy/checks/readability-isolate-declaration.html

It's quite useful for pointers, so I guess it depends how often do you think it could happen. 

One possibility is to fix the initializer but keep the others, and eventually turn off the check and leave it for the reviewers to enforce the rule.

gpanders replied:
I'm strongly in favor of simply disallowing multiple declarations on one line.

justinmk commented on src/nvim/CMakeLists.txt:
```diff
@@ -791,6 +792,22 @@ add_custom_target(
+)
+
+add_custom_target(
+  clang-tidy-full
```
why is this needed? please don't add things just because they're marginally nice-to-have. It increases the surface area that developers have to consume.

justinmk commented on src/nvim/CMakeLists.txt:
```diff
@@ -791,6 +792,22 @@ add_custom_target(
+add_custom_target(
+  clint-non-style
+  COMMAND
+    ${LINT_PRG} --filter=-whitespace,-readability ${LINT_NVIM_REL_SOURCES} 
```
this is not worth adding as a build target, it is just `clint.py` with some arbitrary options.

kylo252 replied:
This does not use the `--supress-errors` flag. I thought it could end beng more performant, but I didn't really benchmark it. Anyways, this is totally disposable. I just wanted to make it clear how to get a clearer baseline.

justinmk commented on src/nvim/CMakeLists.txt:
```diff
@@ -63,6 +63,7 @@ set(LINT_SUPPRESS_FILE ${PROJECT_BINARY_DIR}/errors.json)
 set(LINT_SUPPRESS_URL_BASE "https://raw.githubusercontent.com/neovim/doc/gh-pages/reports/clint")
 set(LINT_SUPPRESS_URL "${LINT_SUPPRESS_URL_BASE}/errors.json")
 set(LINT_PRG ${PROJECT_SOURCE_DIR}/src/clint.py)
+set(CLANG_TIDY_CONFIG_FILE ${PROJECT_SOURCE_DIR}/.clang-tidy)
```
doesn't clang-tidy find this implicitly? that was the reason for putting it in the root dir instead of a subdir like config/

kylo252 replied:
Oh I forgot to point it to the build directory. I also removed the step with linking the database. I'll fix it ASAP.

justinmk commented on Makefile:
```diff
@@ -190,6 +190,18 @@ clint: build/.ran-cmake
+	clang-tidy --config-file .clang-tidy -p build $(FILE)
+
+clang-tidy-full: generate-compilation-db
+	bash -C build clang-tidy-full
```
should this be `$(BUILD_CMD)` ?
```suggestion
	+$(BUILD_CMD) -C build clang-tidy-full
```

justinmk commented on src/nvim/screen.h:
```diff
@@ -59,8 +59,8 @@ extern StlClickDefinition *tab_page_click_defs;
-#define W_ENDCOL(wp)   (wp->w_wincol + wp->w_width)
-#define W_ENDROW(wp)   (wp->w_winrow + wp->w_height)
+#define W_ENDCOL(wp)   ((wp)->w_wincol + (wp)->w_width)
+#define W_ENDROW(wp)   ((wp)->w_winrow + (wp)->w_height)
```
this seems like a good idea

justinmk commented on src/nvim/screen.c:
```diff
@@ -2116,14 +2122,15 @@ static int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, bool noc
-# define WL_SBR         WL_BRI + 1       // 'showbreak' or 'diff'
-#define WL_LINE         WL_SBR + 1      // text in the line
+#define WL_START        0                 // nothing done yet
+# define WL_CMDLINE     (WL_START + 1)    // cmdline window column
```
while we are here, the space in `# define` should be removed in all of these cases. It's an accident from the original Vim source import.

dundargoc replied:
`uncrustify` can fix this. There is a rule that specifically makes the formatting exactly like the style guide (one additional space after # for each level, no space before #).

As for why I haven't fixed it yet...you never asked ;)

kylo252 replied:
I specifically avoided changing that because I knew it's better to fix it in the uncrustify rules :)

kylo252 commented on src/nvim/screen.c:
```diff
@@ -7778,20 +7797,20 @@ void limit_screen_size(void)
-  static long old_Rows = 0;
-  static long old_Columns = 0;
+  static long old_rows = 0;
+  static long old_columns = 0;
```
Yeah this is part of the naming convention [rules](https://github.com/neovim/neovim/pull/15601/files#diff-0534891fbc8b89b4c521057e7e8dff70d3cbeb822cc9991a3abad77848c289f8R35-R36): 
> clang-tidy: Invalid case style for variable 'old_Columns' (fix available) [readability-identifier-naming]

kylo252 commented on src/nvim/syntax.c:
```diff
@@ -6470,8 +6470,7 @@ const char *const highlight_init_cmdline[] = {
   "default link NvimInvalidOptionScope NvimInvalidIdentifierScope",
-  "default link NvimInvalidOptionScopeDelimiter "
-  "NvimInvalidIdentifierScopeDelimiter",
+  "default link NvimInvalidOptionScopeDelimiter NvimInvalidIdentifierScopeDelimiter",
```
@justinmk, is there any chance this could be slightly related to https://github.com/neovim/neovim/issues/10241?

justinmk replied:
no, but thanks for checking :) This is a good change to make. The line-length is not important enough to justify this quirky formatting.

dundargoc commented:
Needs rebase. Also not sure if this if this is that useful. But do a rebase and we can check.

kylo252 committed:
build: additional clang-tidy checks

enable additional useful checks through clangd which were
skipped from `.clang-tidy` for being potentially too noisy

includes rules to that matches Neovim's style-guide:
  - Naming conventions
  - Variable declaration

dundargoc commented on Makefile:
```diff
@@ -148,6 +148,16 @@ distclean:
+clang-tidy-full: generate-compilation-db
+	+$(BUILD_CMD) -C build clang-tidy-full
+
+generate-compilation-db: build/.ran-cmake
+	ninja -C build -t compdb C_COMPILER__nvim_$(CMAKE_BUILD_TYPE) > build/compile_commands.json
```
Is this needed anymore? Doesn't cmake automatically create a compilation database now?

kylo252 replied:
this is specifically for filtering that database, see https://github.com/neovim/neovim/issues/10632#issuecomment-877663416

dundargoc replied:
Wait, that's a completely separate issue. This PR is about introducing clang-tidy, that issue is about "massing" the compilation database for whatever reason.

kylo252 replied:
> for whatever reason

it's specifically for clang gang, see its title :sweat_smile: 
- https://github.com/neovim/neovim/issues/10632

dundargoc commented:
Running `make clang-tidy` gives

```
Error: no input files specified.
USAGE: clang-tidy [options] <source0> [... <sourceN>]
```

dundargoc commented on Makefile:
```diff
@@ -148,6 +148,16 @@ distclean:
 install: checkprefix nvim
 	+$(BUILD_TOOL) -C build install
 
+clang-tidy: generate-compilation-db
```
Running `make clang-tidy` gives 

```
Error: no input files specified.
USAGE: clang-tidy [options] <source0> [... <sourceN>]
```

kylo252 replied:
sorry, this must've broke when I removed the shell script runner. It should be good now.

kylo252 committed:
fixup!: correct fallback to clang-tidy-full

kylo252 committed:
fixup!: remove ninja's compdb override

dundargoc commented on .clangd:
```diff
@@ -1,2 +1,32 @@
+  CompilationDatabase: build
+Diagnostics:
+  # Additional checks that are useful but too noisy to be included in .clang-tidy
+  ClangTidy:
+    Add:
+      - google-*
+      - misc-*
+      - performance-*
+      - readability-*
+    Remove:
+      - misc-no-recursion
+      - misc-unused-parameters
+      - readability-function-*
+      - readability-identifier-length
+      - readability-magic-numbers
+      - readability-redundant-function-ptr-dereference
```
I'm not against this per se, but I can imagine it could annoy other developers that this takes priority over their own configurations. We'll need to be cautious not to add too many things at once I think. I'm not saying "remove this", but just want other people's opinions on this.

dundargoc replied:
Actually, maybe we should separate this and move it to a separate PR? It's not really related to clang-tidy so we don't need to think about this at the moment.

justinmk replied:
this is in a ClangTidy section of .clangd. Does it affect LSP results if I'm using clangd? 

In any case, if this matches our guidelines then it makes sense to have the settings defined in our tree.

It looks like some noisy things are being removed here, which is good.

dundargoc replied:
> Does it affect LSP results if I'm using clangd?

Yes.

OK, we can try it out 👍

dundargoc commented on Makefile:
```diff
@@ -148,6 +148,14 @@ distclean:
 	+$(BUILD_TOOL) -C build install
 
+clang-tidy: build/.ran-cmake
+	@test $(FILE) \
+		&& clang-tidy --config-file .clang-tidy -p build $(FILE) \
+		|| $(MAKE) clang-tidy-full
```
What does `test $(FILE)` even do? It gives the same exit code regardless if the file exists or not. Is there a `-f` missing or something?

Also, I think doing 

```suggestion
	clang-tidy --config-file .clang-tidy -p build $(FILE) || $(MAKE) clang-tidy-full
```

would achieve the same thing.

kylo252 replied:
@dundargoc, did you try it this way?

```sh
make clang-tidy FILE=src/nvim/screen.c
```

dundargoc replied:
Yes?

kylo252 replied:
it's good now either way, since [target has been simplified](https://github.com/neovim/neovim/pull/15601/commits/48609e73aacd1c6a104cf97a3836a2561cd8bbe8)

dundargoc commented on Makefile:
```diff
@@ -148,6 +148,14 @@ distclean:
+		&& clang-tidy --config-file .clang-tidy -p build $(FILE) \
+		|| $(MAKE) clang-tidy-full
+
+clang-tidy-full: build/.ran-cmake
+	+$(BUILD_TOOL) -C build clang-tidy-full
```
You can replace this by adding `clang-tidy-full` to the list with `lintlua lintsh lintuncrustify ...`

dundargoc replied:
Actually, is a make target to run `clang-tidy-full` even needed? There's no way anyone will run this locally since it takes such a long time so I don't think there's much point.

kylo252 replied:
I just wanted to follow the steps of clint, but personally I don't think this specific target offers all that much.

kylo252 committed:
refactor: remove useless target

kylo252 committed:
fix: specify make var

dundargoc commented on .clangd:
```diff
@@ -1,2 +1,32 @@
+      - readability-function-*
+      - readability-identifier-length
+      - readability-magic-numbers
+      - readability-redundant-function-ptr-dereference
```
I still think the warnings are too noisy. I propose we add the following warnings as well.

```suggestion
      - readability-redundant-function-ptr-dereference
      - readability-isolate-declaration
      - bugprone-assignment-in-if-condition
      - readability-suspicious-call-argument
      - performance-no-int-to-ptr
      - readability-else-after-return
      - google-readability-function-size
      - cert-err33-c
```

kylo252 replied:
- `bugprone-assignment-in-if-condition` how often does this even come up and it's not a bug? unless it's some macro
- `cert-err33-c` is pretty serious, see [ERR33-C. Detect and handle standard library errors](https://wiki.sei.cmu.edu/confluence/display/c/ERR33-C.+Detect+and+handle+standard+library+errors)

the rest are style guide related, so I don't mind removing them, although we could only enable them for certain folders/files if they're too noisy for upstream vim files

dundargoc replied:
> how often does this even come up and it's not a bug?

Literally the entire codebase. Did you not test the PR lol?

> cert-err33-c is pretty serious, see [ERR33-C. Detect and handle standard library errors](https://wiki.sei.cmu.edu/confluence/display/c/ERR33-C.+Detect+and+handle+standard+library+errors)

Same thing, it's a valid but it will create massive amounts of noise. This will just result in all of the functions getting a `(void)` slapped on them instead of any error checking.

kylo252 replied:
As discussed, this was a a new check that's only available in v15+ which is why I didn't notice it. I've disabled both for now.

lewis6991 commented on Makefile:
```diff
@@ -148,6 +148,9 @@ distclean:
 	+$(BUILD_TOOL) -C build install
 
+clang-tidy: generated-sources
+	+clang-tidy --config-file .clang-tidy -p build $(FILE)
```
We should define a default value for `FILE`, even if it's empty.

kylo252 replied:
After some testing I figured it's not worth it, since clang-tidy will not run in parallel on its own. We can iterate on this further later on.

kylo252 committed:
fix(tidy): disable new noisy check

dundargoc commented on .clangd:
```diff
@@ -1,2 +1,35 @@
+      - readability-identifier-length
+      - readability-magic-numbers
+      - readability-redundant-function-ptr-dereference
+      - readability-suspicious-call-argument
```
I've already told you that `readability-isolate-declaration` should be removed. Others have told you in [your other PR](https://github.com/neovim/neovim/pull/20446) that they don't like this change. Why haven't you removed it?

kylo252 replied:
oops, sorry! I thought I add it in https://github.com/neovim/neovim/commit/b039593ada338d75aa3f0f80ac66aef6b6656995

kylo252 committed:
fixup!: disable another noisy check

kylo252 committed:
build(tidy): use SEI CERT C recommendations

kylo252 committed:
refactor: remove useless makefile target

this target is useless for multiple files since it runs serially

dundargoc commented on .clangd:
```diff
@@ -1,2 +1,39 @@
+  CompilationDatabase: build
+---
+# Additional useful checks that are too noisy to be included in .clang-tidy
+If:
+  PathMatch: src/nvim/api/.*
```
No, this will make it so that even useful warnings (some of which I rely on) won't show for the regular files since this file takes priority over users own configuration.

This PR is close to being finished, I suggest not making any more changes unless necessary so that we can get this merged. We can iterate on this further in a follow-up PR.

kylo252 replied:
just so we're on the same page, this `If` only restricts the following fragments, this means: 
- src/nvim/api: all configured clang-tidy checks +  these defined here (misc, perf, etc)
- everything else: only use configured clang-tidy checks

<details>
<summary>Demo: notice how the "readability-identifier-naming" is only firing for the file in "src/nvim/api"
</summary>

![image](https://user-images.githubusercontent.com/59826753/195060549-cdc633ed-b344-4ff8-b086-69e41053154e.png)


</details>

given the above,  you still want me to revert this and make these checks global?

dundargoc replied:
Yes, because there are useful warnings in `.clang-tidy` that should be in `.clangd` instead, which are now ignored. Also, I don't understand why we're separating `src/nvim/api` with the rest of the files, shouldn't they have the same configuration?

dundargoc replied:
Do we actually need a `.clang-tidy`? This will be much easier if we just add only necessary warnings to `.clangd` and then add stuff to `.clang-tidy` in a future PR.

kylo252 replied:
> Yes, because there are useful warnings in `.clang-tidy` that should be in `.clangd` instead, which are now ignored.

@dundargoc, this should not be the case, since `.clangd` is not overriding `.clang-tidy` here, just as I explained earlier.

can you please provide an example where that is happening? I've tested on both v14 and v15 and I can't replicate this.

![image](https://user-images.githubusercontent.com/59826753/195108178-1ebd7603-89c2-4c97-a6be-3063cf3bb64b.png)

> Do we actually need a `.clang-tidy`?

yeah we do since clangd doesn't support all the options AFAIK, see https://github.com/clangd/clangd/issues?q=tidy+is%3Aopen

dundargoc replied:
`src/nvim/memline.c` is a good example: there are a bunch of unnecessary casts there and this PR removes those warnings.

kylo252 replied:
oh, okay I think I can see that you mean `readability-non-const-parameter` is useful, where you'd still want it on for all the files.
either way, I'll remove the guard since we can always put it back if it's too noisy.

ii14 replied:
> yeah we do since clangd doesn't support all the options AFAIK, see https://github.com/clangd/clangd/issues?q=tidy+is%3Aopen

@kylo252 what am I supposed to look at here?

dundargoc committed:
build: remove performance-no-int-ptr

dundargoc committed:
build: sort options

dundargoc committed:
build: move everything to .clang-tidy
